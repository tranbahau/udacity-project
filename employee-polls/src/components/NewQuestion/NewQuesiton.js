import { Fragment, useState } from 'react';
import classes from './NewQuesiton.module.css';
import { Button, Form } from 'react-bootstrap';
import { useNavigate } from 'react-router-dom';
import { handleSaveQuestion } from '../../actions/shares';
import { useDispatch, useSelector } from 'react-redux';

function NewQuestion() {
  const navigate = useNavigate();
  const dispatch = useDispatch();
  const authedUser = useSelector((state) => state.authedUser);
  const [option1, setOption1] = useState('');
  const [option2, setOption2] = useState('');

  const handleSave = (e) => {
    e.preventDefault();

    if (option1 === '' || option2 === '') {
      alert('Please fill all option1 or option2');
      return;
    }

    if (option1 === option2) {
      alert('2 option not are the same value');
      return;
    }

    dispatch(
      handleSaveQuestion(
        {
          optionOneText: option1,
          optionTwoText: option2,
          author: authedUser,
        },
        navigate
      )
    );
  };

  const handleCancelClick = () => {
    navigate('/');
  };

  return (
    <Fragment>
      <div className={classes.container}>
        <h3 data-testid="title" className={classes.title}>
          New Question
        </h3>
        <div className={classes.question}>Would you rather?</div>
        <form className={classes.formSubmit} onSubmit={handleSave}>
          <Form.Control
            data-testid="firstOpt"
            as="textarea"
            value={option1}
            className="mb-3"
            onChange={(e) => {
              setOption1(e.target.value);
            }}
            placeholder="Option 1"
          />
          <Form.Control
            data-testid="secondOpt"
            as="textarea"
            value={option2}
            className="mb-3"
            onChange={(e) => {
              setOption2(e.target.value);
            }}
            placeholder="Option 2"
          />
          <div className="d-grid gap-2">
            <Button variant="primary" type="submit" size="sm">
              Save
            </Button>
            <Button variant="secondary" onClick={handleCancelClick} size="sm">
              Cancel
            </Button>
          </div>
        </form>
      </div>
    </Fragment>
  );
}

export default NewQuestion;
