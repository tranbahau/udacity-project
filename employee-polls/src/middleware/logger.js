export const logger = (store) => (next) => (action) => {
  console.group(action.type);
  console.log(action);
  console.groupEnd();
  next(action);
};
