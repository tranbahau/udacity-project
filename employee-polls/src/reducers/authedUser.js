export const authedUser = (state = null, action) => {
  switch (action.type) {
    case 'LOG_OUT':
      return null;
    case 'SET_AUTHED_USER_ID':
      return action.userId;
    default:
      return state;
  }
};
