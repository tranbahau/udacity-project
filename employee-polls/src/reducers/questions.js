export const questions = (state = {}, action) => {
  switch (action.type) {
    case 'GET_QUESTIONS':
      return { ...state, ...action.questions };
    case 'SAVE_QUESTION_ANSWER':
      return {
        ...state,
        [action.answerInfo.questionId]: {
          ...state[action.answerInfo.questionId],
          [action.answerInfo.answer]: {
            ...state[action.answerInfo.questionId][action.answerInfo.answer],
            votes: state[action.answerInfo.questionId][
              action.answerInfo.answer
            ].votes.concat(action.answerInfo.authedUser),
          },
        },
      };
    case 'SAVE_QUESTION':
      return {
        ...state,
        [action.question.id]: action.question,
      };
    default:
      return state;
  }
};
