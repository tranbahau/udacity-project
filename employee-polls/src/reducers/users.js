export const users = (state = {}, action) => {
  switch (action.type) {
    case 'GET_USERS':
      return { ...state, ...action.users };
    case 'SAVE_QUESTION_ANSWER':
      return {
        ...state,
        [action.answerInfo.authedUser]: {
          ...state[action.answerInfo.authedUser],
          answers: {
            ...state[action.answerInfo.authedUser].answers,
            [action.answerInfo.questionId]: action.answerInfo.answer,
          },
        },
      };
    case 'SAVE_QUESTION':
      return {
        ...state,
        [action.question.author]: {
          ...state[action.question.author],
          questions: state[action.question.author].questions.concat(
            action.question.id
          ),
        },
      };
    default:
      return state;
  }
};
