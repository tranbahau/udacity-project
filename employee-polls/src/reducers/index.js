import { combineReducers } from 'redux';
import { users } from './users';
import { questions } from './questions';
import { authedUser } from './authedUser';
import { loadingBarReducer } from 'react-redux-loading-bar';

export const reducers = combineReducers({
  users: users,
  questions: questions,
  authedUser: authedUser,
  loadingBar: loadingBarReducer,
});
