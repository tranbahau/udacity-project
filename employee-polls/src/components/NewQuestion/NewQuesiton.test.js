import { fireEvent, render } from '@testing-library/react';
import { Provider } from 'react-redux';
import { BrowserRouter } from 'react-router-dom';
import NewQuesiton from './NewQuesiton';
import { reducers } from '../../reducers';
import thunk from 'redux-thunk';
import { logger } from '../../middleware/logger';
import { configureStore } from '@reduxjs/toolkit';

describe('New Question Component', () => {
  const store = configureStore({
    reducer: reducers,
    middleware: [thunk, logger],
  });

  it('should render component correctly', () => {
    const component = render(
      <Provider store={store}>
        <BrowserRouter>
          <NewQuesiton />
        </BrowserRouter>
      </Provider>
    );
    expect(component).toBeDefined();
    expect(component).toMatchSnapshot();
  });

  it('should display option text value correctly', () => {
    const component = render(
      <Provider store={store}>
        <BrowserRouter>
          <NewQuesiton />
        </BrowserRouter>
      </Provider>
    );

    const titleElement = component.getByTestId('title');
    const firstOptElement = component.getByTestId('firstOpt');
    const secondOptElement = component.getByTestId('secondOpt');

    expect(titleElement.textContent).toBe('New Question');
    fireEvent.change(firstOptElement, { target: { value: 'React' } });
    fireEvent.change(secondOptElement, { target: { value: 'Angular' } });
    expect(firstOptElement.value).toBe('React');
    expect(secondOptElement.value).toBe('Angular');
  });
});
