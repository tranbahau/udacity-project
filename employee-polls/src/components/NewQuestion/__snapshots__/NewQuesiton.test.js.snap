// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`New Question Component should render component correctly 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="container"
      >
        <h3
          class="title"
          data-testid="title"
        >
          New Question
        </h3>
        <div
          class="question"
        >
          Would you rather?
        </div>
        <form
          class="formSubmit"
        >
          <textarea
            class="mb-3 form-control"
            data-testid="firstOpt"
            placeholder="Option 1"
          />
          <textarea
            class="mb-3 form-control"
            data-testid="secondOpt"
            placeholder="Option 2"
          />
          <div
            class="d-grid gap-2"
          >
            <button
              class="btn btn-primary btn-sm"
              type="submit"
            >
              Save
            </button>
            <button
              class="btn btn-secondary btn-sm"
              type="button"
            >
              Cancel
            </button>
          </div>
        </form>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="container"
    >
      <h3
        class="title"
        data-testid="title"
      >
        New Question
      </h3>
      <div
        class="question"
      >
        Would you rather?
      </div>
      <form
        class="formSubmit"
      >
        <textarea
          class="mb-3 form-control"
          data-testid="firstOpt"
          placeholder="Option 1"
        />
        <textarea
          class="mb-3 form-control"
          data-testid="secondOpt"
          placeholder="Option 2"
        />
        <div
          class="d-grid gap-2"
        >
          <button
            class="btn btn-primary btn-sm"
            type="submit"
          >
            Save
          </button>
          <button
            class="btn btn-secondary btn-sm"
            type="button"
          >
            Cancel
          </button>
        </div>
      </form>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
